hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -489,13 +489,29 @@ public EntityEntry addEntry(
 
 		final EntityEntry e;
 
-		if( (entity instanceof ManagedEntity) &&  ((ManagedEntity) entity).$$_hibernate_getEntityEntry() != null && status == Status.READ_ONLY) {
-			e = ((ManagedEntity) entity).$$_hibernate_getEntityEntry();
-			e.setStatus( status  );
+		/*
+		Important: 		the following instanceof checks and castings are intentional.
+						Please do *not* refactor to make calls through the EntityEntryFactory
+						interface. Refactoring will results in polymorphic call sites which will
+						severely impact performance.
+		*/
+		if (persister.getEntityEntryFactory() instanceof MutableEntityEntryFactory){
+			e = ((MutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
+					status,
+					loadedState,
+					rowId,
+					id,
+					version,
+					lockMode,
+					existsInDatabase,
+					persister,
+					disableVersionIncrement,
+					lazyPropertiesAreUnfetched,
+					this
+			);
 		}
 		else {
-			final EntityEntryFactory entityEntryFactory = persister.getEntityEntryFactory();
-			e = entityEntryFactory.createEntityEntry(
+			e = ((ImmutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
 					status,
```

Comment:
The JVM does need to type check, but before it can type check and cast, it needs to resolve the correct instance using a vtable lookup. The vtable lookup takes 3 bytecode instructions to complete vs 1 bytecode instruction for the direct method call (see https://wiki.openjdk.java.net/display/HotSpot/VirtualCalls).  By casting to a concrete implementation, the JVM does not need to interpret which concrete method to call with each virtual call. We are hard coding that lookup.

This also allows the JVM to perform further optimisations, e.g. inlining method calls.

We can measure the cpu impact using perf, lookig at the method call org/hibernate/event/internal/DefaultLoadEventListener:.loadFromSecondLevelCache()

Before: 28.74% CPU
After: 7.31% CPU